File: src/cpp/factorial.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cstdlib>
#include <chrono>
#include <iomanip>

// Optimized BigInt using Base 10^9
struct BigInt {
    static const int BASE = 1000000000;
    std::vector<int> digits; 

    BigInt(long long n) {
        if (n == 0) digits.push_back(0);
        while (n > 0) {
            digits.push_back(n % BASE);
            n /= BASE;
        }
    }

    void multiply(int n) {
        long long carry = 0;
        for (size_t i = 0; i < digits.size(); ++i) {
            long long current = (long long)digits[i] * n + carry;
            digits[i] = current % BASE;
            carry = current / BASE;
        }
        while (carry) {
            digits.push_back(carry % BASE);
            carry /= BASE;
        }
    }

    void print() const {
        if (digits.empty()) {
            std::cout << "0";
            return;
        }
        std::cout << digits.back();
        for (int i = digits.size() - 2; i >= 0; --i) {
            std::cout << std::setfill('0') << std::setw(9) << digits[i];
        }
    }
};

int main() {
    const char* count_env = std::getenv("COUNT");
    int count = 200;
    if (count_env) {
        count = std::atoi(count_env);
    }

    BigInt factorial(1);
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 2; i <= count; ++i) {
        factorial.multiply(i);
    }
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);

    std::cout << "Result(" << count << "!): ";
    factorial.print();
    std::cout << std::endl;
    std::cout << "Time: " << duration.count() / 1000.0 << " ms" << std::endl;

    return 0;
}

File: src/rust/factorial.rs
use std::env;
use std::time::Instant;

fn main() {
    let count_str = env::var("COUNT").unwrap_or("200".to_string());
    let count = count_str.parse::<u32>().unwrap_or(200);

    let mut digits = vec![1u32];
    
    let start = Instant::now();
    for i in 2..=count {
        let mut carry = 0u64;
        for j in 0..digits.len() {
            let current = digits[j] as u64 * i as u64 + carry;
            digits[j] = (current % 10) as u32;
            carry = current / 10;
        }
        while carry > 0 {
            digits.push((carry % 10) as u32);
            carry /= 10;
        }
    }
    let duration = start.elapsed();

    print!("Result({}!): ", count);
    for &d in digits.iter().rev() {
        print!("{}", d);
    }
    println!();
    println!("Time: {:.3} ms", duration.as_micros() as f64 / 1000.0);
}
